### PR TITLE
Add a DESIGN.md

### PR DESCRIPTION
**Pull request by lxea**
Hello,

This adds an initial design document for each of the features required, as specified in the backend challenge for level 1.

### CODE REVIEW

File: DESIGN.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+## API
+
+The API will be HTTPS with endpoints matching exported functions from
+the library:
+
+- /start - POST - body json `{"command" : "...", "args": [...]}`
+- /stop - POST - query id=...
+- /output - GET - query id=...
+- /status - GET - query id=...
```
Comment (russjones) [inline DESIGN.md:58]: Can you share the request and response objects?

File: DESIGN.md [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+Stop will send a Kill to the associated process, this is a write
+command and as such should lock writes.
+
+---
+
+`Status(ID int) (statusCode, error)`
+
+Status should return a status code to indicate whther the process is
+still running or stopped. This will only require a read lock so
+multiple users can query the process.
```
Comment (russjones) [inline DESIGN.md:40]: What will `statusCode` return?

File: DESIGN.md [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+needed in the future.
+
+The following will be the available interface
+
+---
+
+`Start(command string, args ...string) (ID int, error)`
+
+When Start is called, it will start executing a the program and return
+a unique identifier for the process
```
Comment (russjones) [inline DESIGN.md:17]: How will you generate the unique identifier?

File: DESIGN.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+Once a process is started the logs of the process will be written to a
+file with a name matching the ID. This will allow for easy log
+querying without need for locking a mutex, and reduce memory needed
+for noisy logs.
```
Comment (russjones) [inline DESIGN.md:22]: What will the logs be?

File: DESIGN.md [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+the HTTP API service.
+
+### Authorization & Authentication
+
+The HTTP API will implement authentication, here HTTP Basic auth will
+be use and a preset list of users will be hard-coded in this case.
+
+Once a user is authenticated they will only be authorized to access or
+modify processes started by that same user, any unauthorized access
+will be logged and an unauthorized HTTP error code will be returned.
```
Comment (russjones) [inline DESIGN.md:75]: This is fine.

File: DESIGN.md [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+output and JSON would make the
+
+The API will be HTTPS with endpoints matching the libraries exposed
+functions, however there will be additional authentication handled by
+the HTTP API service.
+
+### Authorization & Authentication
+
+The HTTP API will implement authentication, here HTTP Basic auth will
+be use and a preset list of users will be hard-coded in this case.
```
Comment (russjones) [inline DESIGN.md:71]: Can you provide some more details here?

File: DESIGN.md (hunk 4 above)
Comment (lxea) [reply DESIGN.md:22]: Not sure what you mean here -- I mean for them to be the stdout of process

Comment (russjones) [reply DESIGN.md:24]: Any reason you're not capturing stderr?

Comment (lxea) [reply DESIGN.md:24]: No, I'll update the api/library sections to mention this/access

File: DESIGN.md [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+---
+
+`Status(ID int) (statusCode, error)`
+
+Status should return a status code to indicate whther the process is
+still running or stopped. This will only require a read lock so
+multiple users can query the process.
+
+`statusCode` will be an constant representing either a RUNNING or
+STOPPED status:
```
Comment (timothyb89) [inline DESIGN.md:45]: Does `STOPPED` cover all the various ways the process could end, or just some of them?

Comment (lxea) [reply DESIGN.md:45]: `STOPPED` here is just a catch all that the program ended, it wont check how/why -- if the exit code wasnt 0 etc

File: DESIGN.md [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+The following will be the available interface
+
+---
+
+`Start(command string, args ...string) (ID int, error)`
+
+When Start is called, it will start executing a the program and return
+a unique identifier for the process. The unique identifier will be
+generated by incrementing an integer upon each time a service
+starts.
```
Comment (russjones) [inline DESIGN.md:19]: You would be leaking information about the existence of jobs here, what do you think about just using a UUID?

That way your ID would be random enough it's unguessable.

File: DESIGN.md [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+  - *response*: `{"id": ...}`
+- /stop - POST
+  - *request query*: id=...
+  - *response* -- only if there is a non 200 response, `{ "error": ... }`
+- /output - GET
+  - *request query*:
+	  - id=..., x
+	  - Optional query `stream` with options `stdout` and `stderr`,
+        default to stdout
+  - *response*: non JSON formatted log output
```
Comment (russjones) [inline DESIGN.md:80]: Why not just return JSON so your API is consistent?

File: DESIGN.md [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+- /start - POST
+  - *request body*: `{"command" : "...", "args": [...]}`
+  - *response*: `{"id": ...}`
+- /stop - POST
+  - *request query*: id=...
+  - *response* -- only if there is a non 200 response, `{ "error": ... }`
+- /output - GET
+  - *request query*:
+	  - id=..., x
+	  - Optional query `stream` with options `stdout` and `stderr`,
```
[the remark this answers is not in the document]
Comment (lxea) [reply DESIGN.md:78]: Yes exactly

File: DESIGN.md [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+
+Status should return a status code to indicate whther the process is
+still running or stopped. This will only require a read lock so
+multiple users can query the process.
+
+`statusCode` will be an constant representing either a RUNNING or
+STOPPED status:
+
+```go
+type StatusCode int
```
[the remark this answers is not in the document]
Comment (lxea) [reply DESIGN.md:49]: This was intended so I could stringify the enum more easily, not as the program exit code, but I think its worth also including the exit code separately

File: DESIGN.md (hunk 7 above)
Comment (lxea) [reply DESIGN.md:46]: I've added an additional `Killed` enum entry to represent this case as well as also returning the programs exit code

**Review by russjones — APPROVED** [review]

**Review by timothyb89 — APPROVED** [review]

**Pull request merged**